### ISSUE TITLE
Columns from queries are suddenly nullable since Drift 2.19

### ISSUE THREAD

**Issue by frankvollebregt** [simolus3/moor#3101]
**Describe the bug**
I assume it has to do with this change from the changelog:
> Drift files: Make columns coming from subquery expressions nullable.

In my `.drift` file, I have something like this defined:
```sql
myQuery: SELECT
  ...
  (SELECT IFNULL(CAST(SUM(other_table .some_number) AS INT), 0) FROM other_table WHERE ...)
FROM some_table
```
In this case, I would expect `drift_dev` to realise that the result of this subquery can never be `null`. If I add an explicit extra `CAST(... AS INT)` around the whole thing, it still behaves the same.

Is this just something that is _very_ difficult to determine at build time? Or is this expected to work and return a non-nullable `int` as its result?

## Edit
I also encounter this in another subquery which looks like:
```sql
SELECT
  ...
  (SELECT COUNT(*) = 0 FROM ... WHERE ...)
```
Which returns a `bool?` instead of a `bool`

**Comment by simolus3** [comment]
In older versions, `drift_dev` didn't consider that subqueries could select from an empty table at all and assumed that a subquery `(SELECT <e> FROM ...)` is non-null if `<e>` is non-null. This is caused instances where we insert null assertions for values that can be null (because the table is empty and the query is not an aggregate).

I've tried to avoid false-positives here by not considering the subquery to be nullable if the expression is an aggregate function call. Looks like we also need to do this if any subexpression is aggregate as long as the table doesn't have a `GROUP BY` clause.

**Comment by simolus3** [comment]
This should be fixed with the linked commit. I didn't test your exact queries because I don't  have the tables. If you have the time, could you check whether it works for you with that commit? You may have to run `dart run build_runner clean` so that the build-script is invalidated.

```yaml
dependency_overrides:
  sqlparser:
    git:
      url: https://github.com/simolus3/drift.git
      ref: develop
      path: sqlparser
```

If that fixes it I'll do a patch release for `sqlparser`.

**Comment by frankvollebregt** [comment]
I'll try it out, thank you for looking into this so quickly

**Comment by frankvollebregt** [comment]
This version fixes some of the errors, but there are two left:
1. A case with `IFNULL` and `MIN`, which looks like this:
```sql
SELECT
  ...,
  (
    SELECT IFNULL(MIN(r_actions.status), 0) FROM actions
    JOIN r_actions USING (r_actionsid)
    WHERE parent_id = :id
  ) AS minStatus
FROM ...
```
2. A case with `IFNULL`, `CAST AS INT` and `SUM`:
```sql
SELECT
  ...,
  (
    SELECT IFNULL(CAST(SUM(orderlines.amount) AS INT), 0) FROM orders
    JOIN orderlines USING (order_id)
    WHERE
      ... -- some complex where condition (shouldn't matter I guess)
  ) AS total_amount
```
The full data model is a bit complex and contains some localized column names, so I don't think it'll be of much help. Hopefully this can help track down the remaining issues.

### Cases it did resolve
- Cases with `SELECT COUNT(*) = 0`, like the example I shared, now return a `bool`
- One with a pretty complex `CASE` statement, comparing various columns and using `CAST`/`SUM`, before multiplying the result of this `CASE` with a number such as `-1`

**Comment by simolus3** [comment]
Thanks for the follow-up, the problem was that we didn't consider aggregate-function arguments to non-aggregate functions. I've fixed that problem as well and released a fix with `sqlparser` version `0.37.1`.

**Comment by frankvollebregt** [comment]
Awesome! Everything works as expected now, thank you!
